📝 Update structure in development docs

tiangolo opened the pull request:
<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?

This is a small update for the structure of the docs about building Ray from source.

My idea was to isolate steps that are shared and then steps required per platform/system. Also consolidating the instructions to clone with git, install, directory structure, etc.

I'm still figuring out the building steps (installing the dependencies for docs in an M1), but I wanted to start the draft right away.

<!-- Please give a short summary of the change and the problem this solves. -->

## Related issue number

<!-- For example: "Closes #1234" -->

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

pcmoritz commented on doc/source/ray-contribute/development.rst:
```diff
@@ -10,6 +10,68 @@ For a majority of Ray users, installing Ray via the latest wheels or pip package
+
+.. code-block:: shell
+
+    python3.8 -m venv env
```
Would it be better to do `python -m venv env` here? That should work across Python versions, right?

tiangolo replied:
You're right! I also updated a comment below with an example for Python 3.8 only, to clarify that it's an example and not the only version supported.

tiangolo commented:
I think this is ready for review! I see some errors in CI, but they don't seem related to my changes. 🤔

maxpumperla commented on doc/source/ray-contribute/development.rst:
```diff
@@ -10,6 +10,68 @@ For a majority of Ray users, installing Ray via the latest wheels or pip package
+
+(Optional) To setup an isolated Anaconda environment, see :ref:`ray_anaconda`.
+
+In any case, you will probably want to have some type of Python virtual environment.
```
Maybe we shouldn't make every sentence a new paragraph? I think this renders / reads not that great.

tiangolo replied:
Sure! I merged this. I think this also applies to other places so I merged other paragraphs too. 🤓

maxpumperla commented on doc/source/ray-contribute/development.rst:
```diff
@@ -10,6 +10,68 @@ For a majority of Ray users, installing Ray via the latest wheels or pip package
+Prepare the Python environment
+------------------------------
+
+(Optional) To setup an isolated Anaconda environment, see :ref:`ray_anaconda`.
```
Can we please make this a `.. tip:`, instead of `(Optional)`?

tiangolo replied:
Sure thing!

maxpumperla commented on doc/source/ray-contribute/development.rst:
```diff
@@ -74,39 +133,55 @@ For RHELv8 (Redhat EL 8.0-64 Minimal), run the following commands:
 
-Install bazel manually from link: https://docs.bazel.build/versions/main/install-redhat.html 
-
+In RedHat, install bazel manually from this link: https://docs.bazel.build/versions/main/install-redhat.html
```
Shall we try to keep Bazel uppercase consistently?

tiangolo replied:
I like that! I did the same for Brew too.

maxpumperla commented on doc/source/ray-contribute/development.rst:
```diff
@@ -45,27 +101,30 @@ Running the script will remove the  ``ray/tune``, ``ray/rllib``, ``ray/autoscale
 
 .. tip:: If you are only editing Tune/RLlib/Autoscaler files, follow instructions for :ref:`python-develop` to avoid long build times.
 
-To build Ray, first install the following dependencies.
```
I find "the following" to be clearer. Maybe settle for:

`To build Ray on Ubuntu, run the following commands:`

to avoid the clumsy duplication of the word "following" (if that's the concern).

tiangolo replied:
Sure thing! Done.

maxpumperla commented on doc/source/ray-contribute/development.rst:
```diff
@@ -74,39 +133,55 @@ For RHELv8 (Redhat EL 8.0-64 Minimal), run the following commands:
-  ray/ci/env/install-bazel.sh
-  # (Windows users: please manually place Bazel in your PATH, and point
-  # BAZEL_SH to MSYS2's Bash: ``set BAZEL_SH=C:\Program Files\Git\bin\bash.exe``)
+You will also need to install NodeJS to build the dashboard: https://nodejs.org
```
Can we make `NodeJS` a link to the URL and not have the latter as plain text? thanks!

maxpumperla replied:
Also, in this case I'd again try to consolidate a couple of sentences into a paragraph for a calmer reading experience.

tiangolo replied:
Perfect, done! 🚀

maxpumperla commented on doc/source/ray-contribute/development.rst:
```diff
@@ -74,39 +133,55 @@ For RHELv8 (Redhat EL 8.0-64 Minimal), run the following commands:
+------------------------------------
 
-  git clone https://github.com/ray-project/ray.git
+Make sure you have a local clone of the git repository as explained above.
```
Can we specify the we mean the Ray source repository by that?

tiangolo replied:
Yep. ✅

maxpumperla commented on doc/source/ray-contribute/development.rst:
```diff
@@ -74,39 +133,55 @@ For RHELv8 (Redhat EL 8.0-64 Minimal), run the following commands:
+
+Make sure you activate any Python virtual (or conda) environment you could be using as described above.
+
+Enter into the directory for Python inside of the Ray project directory and install the project with ``pip``:
```
maybe say `python` here for clarity (specifying that this is indeed the folder name you mean)

tiangolo replied:
Good point! I used `python/` to make it explicit that I'm referring to the directory.

maxpumperla commented on doc/source/ray-contribute/development.rst:
```diff
@@ -170,7 +254,7 @@ Define an environment variable BAZEL_PATH to full exe path (example:
 --------------------------------------------
 
-You can tweak the build with the following environment variables (when running ``setup.py``):
+You can tweak the build with the following environment variables (when running ``setup.py`` or ``pip install -e .``):
```
in contrast to the new command, maybe we should now say `python setup.py install`?

tiangolo replied:
Agreed! 🤓

maxpumperla commented on doc/source/ray-contribute/development.rst:
```diff
@@ -248,11 +332,43 @@ update the documentation!
 When you make documentation changes, build them locally to verify they render
 correctly. `Sphinx <http://sphinx-doc.org/>`_ is used to generate the documentation.
 
+Building Docs for Apple Silicon (M1)
```
Quick question: Is it possible to entirely move / consolidate the "Building the Docs" section into `ray-contribute/docs.rst`? I think this might just be a little redundant (not necessarily a bad thing, but this document is already super long!)

tiangolo replied:
I hadn't seen that, I see that one is a Jupyter Notebook. Let me try to move those parts there.

tiangolo replied:
I think I got it right. 🤓  I moved all the docs about building the docs to that section/file and made a note on the main development page pointing to that one.

maxpumperla replied:
awesome, thank you!

maxpumperla reviewed (changes requested):
@tiangolo thanks so much for this amazing PR! Huge fan of your work, btw. 🥇

I have a couple of minor nits, but otherwise we're good to go. :v:

tiangolo commented on doc/source/ray-contribute/development.rst:
```diff
@@ -10,6 +10,68 @@ For a majority of Ray users, installing Ray via the latest wheels or pip package
+
+.. code-block:: shell
+
+    git clone https://github.com/[your username]/ray.git
```
Cool! I'm not sure what you mean with a *code tab*, is there a particular syntax for that? Or would it be just another block like this underneath?

Also, I think this one is the https version, right? I guess the other one would be the SSH one, with like:

```console
$ git clone  git@github.com:[your username]/ray.git
```

Right? Or do you mean something else?

pcmoritz replied:
I think there is an example of this in `walkthrough.rst`, you should be able to do it like this:

```
Installation
------------

.. tabbed:: Python

    To run this walkthrough, install Ray with ``pip install -U ray``. For the latest wheels (for a snapshot of ``master``), you can use these instructions at :ref:`install-nightlies`.

.. tabbed:: Java

    To run this walkthrough, add `Ray API <https://mvnrepository.com/artifact/io.ray/ray-api>`_ and `Ray Runtime <https://mvnrepository.com/artifact/io.ray/ray-runtime>`_ as dependencies. Snapshot versions can be found in `sonatype repository <https://oss.sonatype.org/#nexus-search;quick~io.ray>`_.

    Note: To run your Ray Java application, you need to install Ray Python with `pip install -U ray` first. (For Ray Java snapshot versions, install nightly Ray Python wheels.) The versions of Ray Java and Ray Python must match.

.. tabbed:: C++

[...]
```

On https vs. ssh, I think that's what Max meant yeah :)

pcmoritz replied:
See the rendering here: https://docs.ray.io/en/latest/ray-core/walkthrough.html

tiangolo replied:
Awesome, thanks! Done. ✅

tiangolo commented on doc/source/ray-contribute/development.rst:
```diff
@@ -19,24 +81,18 @@ Building Ray (Python Only)
+    pip install -U https://s3-us-west-2.amazonaws.com/ray-wheels/latest/ray-2.0.0.dev0-cp38-cp38-manylinux2014_x86_64.whl
 
-3. Replace Python files in the installed package with your local editable copy. We provide a simple script to help you do this: ``python python/ray/setup-dev.py``.
+4. Replace Python files in the installed package with your local editable copy. We provide a simple script to help you do this: ``python python/ray/setup-dev.py``.
```
Thanks! I think I fixed.

tiangolo commented on doc/source/ray-contribute/development.rst:
```diff
@@ -248,11 +332,43 @@ update the documentation!
+
+.. code-block:: shell
+
+  conda install -c conda-forge xboost lightgbm
```
Thanks! ✅

tiangolo commented:
Thanks a lot for all the feedback @maxpumperla! 🚀 

I implemented it all and I moved the info about building docs to the article dedicated to building docs (a Jupyter Notebook).

pcmoritz reviewed (approved):
Thanks a lot, this is really great ❤️ ! It will make it much easier for people to contribute :)

I'll leave the final approval to @maxpumperla, thank you both for making this happen!

Small side node:

I wonder if we should do the same tabbing for the `git remote add upstream https://github.com/ray-project/ray.git` step also?

tiangolo commented:
So good to hear this seems useful @pcmoritz! 🚀 

I like your idea, I added tabbed instructions to set up the remote with SSH and HTTPS. 🤓

maxpumperla commented on doc/source/ray-contribute/development.rst:
```diff
@@ -74,39 +133,55 @@ For RHELv8 (Redhat EL 8.0-64 Minimal), run the following commands:
 
-  # Build the dashboard
-  # (requires Node.js, see https://nodejs.org/ for more information).
-  pushd ray/dashboard/client
```
yeah, removing is better! thanks
